Fix --ignore-imports for multi-line imports

chkno opened the pull request:
This fixes #1422 and fixes #2019.

(I tried running the tox tests, but some of them seemed to be failing at HEAD, and I didn't have all the interpreters that it wanted to check installed on my machine.)

matthewfranglen commented on pylint/checkers/similar.py:
```diff
@@ -137,10 +138,16 @@ def stripped_lines(lines, ignore_comments, ignore_docstrings, ignore_imports):
+        tree = ast.fix_missing_locations(ast.parse(''.join(lines)))
+        line_is_import = {
+            node.lineno: isinstance(node, (ast.Import, ast.ImportFrom))
+            for node in tree.body}
```
This results in the final expression of the line overwriting any previous ones. For example:

```python
from foo import bar; print (bar)
```

This results in a `line_is_import` of:
```
{1: False}
```

It's horrible code, however it is legal python. Something like this may work:

```python
expression_is_import = [
    (node.lineno, isinstance(node, (ast.Import, ast.ImportFrom)))
    for node in tree.body]
line_is_import = {
    lineno: any(is_import for _, is_import in lineno_is_import)
    for lineno, lineno_is_import in groupby(expression_is_import, key=lambda expression: expression[0])}
```

chkno replied:
I see what you mean, but I think `any` should be `all`: --ignore-imports should only ignore imports; the presence of any code should keep the line under consideration.

PCManticore commented on pylint/checkers/similar.py:
```diff
@@ -18,8 +18,10 @@
 """
 
 from __future__ import print_function
+import ast
```
Let's replace this with `astroid` instead. The API is similar with the `ast` module so you shouldn't need to modify anything other than the import.

chkno replied:
Done.

PCManticore commented on pylint/checkers/similar.py:
```diff
@@ -152,7 +164,9 @@ def stripped_lines(lines, ignore_comments, ignore_docstrings, ignore_imports):
                 line = ''
         if ignore_imports:
-            if line.startswith("import ") or line.startswith("from "):
+            if lineno in line_is_import:
```
These two lines could be rewritten as `current_line_is_import = line_is_import.get(lineno)`

chkno replied:
It can't:  When `lineno` is not in `line_is_import`, `current_line_is_import` needs to keep its previous true/false value.  Many lines are not present in line_is_import.  I those cases, we must continue to use the value from the previous line.

`current_line_is_import = line_is_import.get(lineno, current_line_is_import)` would work, but this doesn't read as clearly to me.

I renamed `line_is_import` to `line_begins_import` to try to clarify this.

PCManticore replied:
Oh gotcha, this makes sense then. But I definitely suggest squashing this into `line_is_import.get(lineno, current_line_is_import)` as there is no reason to do the lookup and the retrieval separately.

chkno replied:
Um, ok.  Done.

(If you don't mind making this a teaching moment: Why?  I thought it might be for performance, but I measured and found that using .get() is slower (p=0.002, 95% CI slower by 130 - 570 ns/line) in cpython 3.6 (it trades a LOAD_FAST, COMPARE_OP, POP_JUMP_IF_FALSE, and BINARY_SUBSCR for a LOAD_ATTR, CALL_FUNCTION, and whatever the function does).  Is it expected to be faster in pypy and people who care about performance run pylint with pypy, or is this just a python idiom I need to absorb?  Thanks!)

PCManticore replied:
Oh, to be honest I was under the impression that this would have been faster than a lookup plus an additional access, put apparently that's not the case! So instead if was a teaching moment for me. :) Nonetheless, reducing the two lines to one looks better to me.

PCManticore commented on pylint/test/unittest_checker_similar.py:
```diff
@@ -90,6 +94,51 @@ def test_ignore_imports():
 
 
+def test_multiline_imports():
+    sys.stdout = StringIO()
```
This idiom seems like a perfect candidate for `contextlib. redirect_stdout`

chkno replied:
Yes, it does.  I'd like to keep this pull request narrowly focused on --ignore-imports, so I sent a separate pull request with this change: https://github.com/PyCQA/pylint/pull/2432

chkno commented:
I updated Changelog and CONTRIBUTORS.  (I saw these instructions in doc/development_guide/contribute.rst, but I wasn't sure if they were intended for a change this small.  Thank you for walking me through this.)

(I rebased to resolve a Changelog merge conflict.  Sorry if this was the wrong thing to do.)

PCManticore commented:
Thanks a lot for the PR @chkno ! I've merged this manually to solve the conflicts and just pushed the relevant commit to `master`.
